### DISCUSSION TITLE
Excessive memory usage

### DISCUSSION THREAD

**Discussion by tuimz** [spatie/laravel-permission#2087] [Q&A]
**Describe the bug**
We have 188 permissions, 8 roles, 182 records in role_has_permissions and 32 records in model_has_permissions our database table. 

Take the following test request in `web.php`

```php
Route::get('/test', function () {
    return response('OK', 200);
})
```

Wich results in about 40mb memory usage. However if we change that to

```php
Route::get('/test', function () {
    auth()->user()->can('view-dashboard');
    return response('OK', 200);
})
```
Memory used for the request to ~530mb

**Versions**
- spatie/laravel-permission package version: 5.5.2
- illuminate/framework package: 8.83.4

PHP version: 8.1.3 

Database version: Mysql 8.0.28

**Environment (please complete the following information, because it helps us investigate better):**
 - OS: Linux
 - Version: Kernel 5.16.11

**Comment by tuimz** [comment]
I want to add if I set the config `permission.cache.store` to  `array` the memory usage goes down to 40mb again

**Comment by erikn69** [comment]
I have 389 permissions and 13 roles and many records on `role_has_permissions`/ `model_has_permissions`, and only 6mb with laravel debugbar, without debugbar 3 or 4, you are doing something wrong, test on a fresh instalation
![image](https://user-images.githubusercontent.com/4933954/158610451-6a26709f-a82c-45eb-bdf4-e0d524618709.png)

**Comment by salvorl** [comment]
@tuimz did you solve it? 
I've a similar problems, and I don't know how manage it.
I have 6694 Role and 137 Permission.

**Reply by parallels999** [reply]
> I have 6694 Role

really a lot of roles, you're supposed to have lots of permissions and few roles
make a debug for find the problem, and fell free to make a pull request fixing it
Also, wich version are you using?

**Reply by salvorl** [reply]
The 6694 roles are loaded by roles_has_permission table, in the roles table I've 131 row.
![image](https://user-images.githubusercontent.com/99655111/211500193-4de4cec5-1116-4af5-949e-99a7507e293a.png)


My debugbar loads model by spatie ( in erikn69 reply image the roles are in App/model). 
I've to check if there is a configuration problem.
That if role cache is set to "array", and the memory used iss 33MB

Using role cache to "default" the memory used is 110 MB and the spatie/permission/model/role are not loaded
![image](https://user-images.githubusercontent.com/99655111/211500021-4c68a5dc-f14e-4598-8d7a-5b2e57ddd352.png)


My php version is 7.1 and spatie version is 2.37

**Reply by parallels999** [reply]
> spatie version is 2.37

spatie current version is **5.7**, Most likely this problem has already been fixed.

- https://github.com/spatie/laravel-permission/pull/1799
- https://github.com/spatie/laravel-permission/pull/1912

**Reply by erikn69** [reply]
> My debugbar loads model by spatie in erikn69 reply image the roles are in App/model).

@salvorl I have custom models extending spatie models 

>spatie version is 2.37

Spatie laravel permissions 2.37 has no support, only latest version(V5) has support

**Reply by salvorl** [reply]
The project was at laravel 5.8.
Now I'm trying to update to laravel 6, after that I'll update to laravel 7.

PS
I want to report that in the prerequisite of V5 documentation the laravel version supported is 6, in last composer is 7

**Reply by salvorl** [reply]
With the 5.7 version the improvement is impressive. 
The memory go down from 110 MB to 9 MB

**Reply by erikn69** [reply]
> I want to report that in the prerequisite of V5 documentation the laravel version supported is 6, in last composer is 7

In the first v5 releases, it supported laravel 6

> With the 5.7 version the improvement is impressive.

Thanks,  it was one of my first contributions #1799

Also this idea could help with memory and time https://github.com/spatie/laravel-permission/pull/1833.
